soc/arm/atmel_sam0/common/soc_saml2x.c

```diff
@@ -0,0 +1,272 @@
+		|  GCLK_PCHCTRL_CHEN
+		|  GCLK_PCHCTRL_GEN_GCLK1;
+#endif
+
+	/* --- */
+
+	/* if the target frequency is 48 MHz, then the calibration value can be used to
+	 * decrease the time until the coarse lock is acquired. this is loaded from
+	 * NVMCTRL_OTP5[31:26]
+	 */
+	cal = *((uint32_t *)NVMCTRL_OTP5);
+	cal >>= 26;
+	cal &= (1 << 6) - 1;
+
+	OSCCTRL->DFLLCTRL.reg = 0
+		|  OSCCTRL_DFLLCTRL_QLDIS
+		| !OSCCTRL_DFLLCTRL_ONDEMAND
+		|  OSCCTRL_DFLLCTRL_RUNSTDBY
+#if !CONFIG_SOC_ATMEL_SAML_OPENLOOP_AS_MAIN
+		|  OSCCTRL_DFLLCTRL_MODE
+#endif
+		;
+
+	OSCCTRL->DFLLVAL.reg = 0
+		|  OSCCTRL_DFLLVAL_COARSE(cal)
+		|  OSCCTRL_DFLLVAL_FINE(512) /* use 50% */
+		;
+
+	OSCCTRL->DFLLMUL.reg = 0
+		/* use 25% of maximum value for the coarse and fine step
+		 * ... I couldn't find details on the inner workings of the DFLL, or any
+		 * example values for these - I have seen others using ~50%. hopefully these
+		 * values will provide a good balance between startup time and overshoot
+		 */
+		|  OSCCTRL_DFLLMUL_CSTEP(16)
+		|  OSCCTRL_DFLLMUL_FSTEP(256)
+#if CONFIG_SOC_ATMEL_SAML_OSC32K_AS_MAIN || CONFIG_SOC_ATMEL_SAML_XOSC32K_AS_MAIN
+		/* use a 32.768 kHz reference ... 48e6 / 32,768 = 1,464.843... */
+		|  OSCCTRL_DFLLMUL_MUL(1465)
+#elif CONFIG_SOC_ATMEL_SAML_OSC16M_AS_MAIN
+		/* use a 16 MHz -> 31.25 kHz reference... 48e6 / 31,250 = 1,536
+		 * a small value can make the DFLL unstable, hence not using the
+		 * 16 MHz source directly
+		 */
+		|  OSCCTRL_DFLLMUL_MUL(1536)
+#endif
+		;
+
+	/* --- */
+
+	/* enable */
+	while (!OSCCTRL->STATUS.bit.DFLLRDY) {
+	}
+	OSCCTRL->DFLLCTRL.bit.ENABLE = 1;
+
+#if !CONFIG_SOC_ATMEL_SAML_OPENLOOP_AS_MAIN
+	/* wait for ready... note in open loop mode, we won't get a lock */
+	while (!OSCCTRL->STATUS.bit.DFLLLCKC || !OSCCTRL->STATUS.bit.DFLLLCKF) {
+	}
+#endif
+}
+
+static inline void flash_waitstates_init(void)
+{
+	/* PL2, >= 2.7v, 48MHz = 2 wait states */
+	NVMCTRL->CTRLB.bit.RWS = 2;
+}
+
+static inline void pm_init(void)
+{
+	PM->PLCFG.bit.PLDIS = 0;
+	PM->PLCFG.bit.PLSEL = 2;
+}
+
+static inline void gclk_main_configure(void)
+{
+	/* finally, switch the CPU over to run from DFLL48M */
+	GCLK->GENCTRL[0].bit.SRC = GCLK_GENCTRL_SRC_DFLL48M_Val;
+}
+
+#if !CONFIG_ADC_SAM0
+#define gclk_adc_configure()
+#else
+static inline void gclk_adc_configure(void)
+{
+	GCLK->GENCTRL[3].reg = 0
+		| GCLK_GENCTRL_SRC_DFLL48M
+		| GCLK_GENCTRL_DIV(2)
+		| GCLK_GENCTRL_GENEN;
+}
+#endif
+
+void pause_for_debug(void)
+{
+	/* for some reason, when attempting to flash / debug the target, the operations
+	 * will time out... I suspect this is due to clock configuration, so instead of
+	 * doing this immediately, we defer startup for a while to permit the debugger
+	 * to jump in and interrupt us. ick
+	 */
+	for (uint32_t i = 0; i < 10000; i += 1) {
+		__asm__ volatile ("nop\n");
+	}
+}
+
+static int atmel_saml_init(const struct device *arg)
+{
+	uint32_t key;
+
+	ARG_UNUSED(arg);
+
+	key = irq_lock();
+
+	pause_for_debug();
```

Comment:
My feeling is this should be at least at Kconfig. It could be dropped when you produce binary in release mode too.

Reply:
Well caught... I was actually going to ask if anyone else has seen this sort of behaviour. (I guess not by its absence in other SoCs...)

I'll make it a config option, though to if you agree, I'll make this `default y`, otherwise people will have a terrible time getting things up and running.

Reply:
See [here](https://github.com/zephyrproject-rtos/zephyr/pull/41308/commits/e9548d55363040712af337a088c7a087b67d6c3d#diff-1630f78dc46dcfb3dc029e1d30e8e2e35826e0a4b49dcec575facb376a0c6f90R6-R16) for commentary.